src/Control/Middleware/RegisteredEventsMiddleware.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace SilverStripe\Control\Middleware;
+
+use InvalidArgumentException;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Injector\Injectable;
+
+/**
+ * Allows events to be registered and passed through middleware.
+ * Useful for event registered prior to the beginning of a middleware chain.
+ */
+class RegisteredEventsMiddleware implements HTTPMiddleware
```

Comment:
I would have expected a DomainCanoncialisationMiddleware instead of this more abstract one?

Reply:
And then a `DomainCanoncialisationMiddleware.force_www` config option or something...

Reply:
The thing that weirds me out about this a bit is that a "registered event" and a "middleware" are two different abstractions that do very different things, in addition to the legacy "request filter" abstraction.

Reply:
The alternative approach would be to turn the forceWWW logic into its own middleware, and then add that middleware to Director instead of registering an event.